Object Recognition

jcarnegieGN opened the pull request:
This implements the object recognition in papers:
C. Papazov and D. Burschka An Efficient RANSAC for 3D Object Recognition in Noisy and Occluded Scenes
C. Papazov, S. Haddadin, S. Parusel ,K. Krieger, D. Burschka "Rigid 3D geometry matching for grasping of known objects in cluttered scenes."

It finds the objects in a scene that are in a database and a transform which aligns the model to the scene. I still need to test recognition. The descriptor used is one huge database of point pair features with the local coordinate system of each pair described in a homogeneous matrix. The size of it when I used it on the test data base used in the Ensemble of Shape Functions paper was several gigabytes, so the descriptor creation might be changed to only include the points.

ruffsl commented:
@jcarnegieGN , big thanks for the PR, and for contributing back to OmniMapper! Could you add a README.md to the fsr_3d_object_recognition directrory with some detials on the implementation. Also does this codebase come with any given licence or origin? OmniMapper is releases under the BSD licence.

Just a note, I won't have the chance to review this in full until I get back to my workstation at GT in late August.

jcarnegieGN commented:
I added a new README.md. The details aren't too specific because the details of the implementation may change. The codebase comes with a BSD license, sorry, I should've added that. I'll add the license in the readme. The recognition isn't working yet, I'll add omnimapper specific code after I've gotten it working.

As far as progress, I've switched from using partial views to using full 3d models for object description in order to save memory for the object database. Using partial views meant there was an exorbitant amount of redundant information per model in the database. The relative scale of models between description creation and use of the description was an issue, so the recognizer scales the model according to the resolution of the scene before it matches it. I now have a problem which I believe has to do with pcl using a viewpoint to compute the normals of a cloud. I'm going to try computing the normals for the model in Meshlab and then saving them with the normals computed. Then when I create the point clouds in vtk I should be able to solve the normals in a pcd file.
